ui/src/features/Editor/components/LeftPanel/index.tsx

```diff
@@ -40,21 +41,63 @@ const LeftPanel: React.FC<Props> = ({
   onNodesAdd,
   isMainWorkflow,
   hasReader,
-  // onNodeDoubleClick,
-  // selected,
+  onNodesChange,
+  onNodeDoubleClick,
+  selected,
 }) => {
   const t = useT();
   const { workspaceId } = useParams({ strict: false });
   const [selectedTab, setSelectedTab] = useState<Tab | undefined>();
-  const { fitView } = useReactFlow();
+  const { fitView, getViewport, zoomTo } = useReactFlow();
   const [nodeId, setNodeId] = useState<string | undefined>(undefined);
 
-  useEffect(() => {
+  const previousZoomRef = useRef<number | undefined>(undefined);
+
+  const handleSelectionChange = useCallback(() => {
```

Comment:
If this is only being used inside the useEffect, just move the code there.

Also, please review the contents. Currently, at the very least, you will each time be doing setNodeId(undefined) twice, which is obviously not good. Please make sure each conditional makes 100% sense to you and is not redundant 🙏🏻